File: pkg/orchestrator/shutdown.go
package orchestrator

import (
	"context"
	"fmt"
	"os"
	"os/signal"
	"sync"
	"syscall"
	"time"

	"github.com/billm/baaaht/orchestrator/internal/logger"
	"github.com/billm/baaaht/orchestrator/pkg/types"
)

// ShutdownState represents the current state of the shutdown process
type ShutdownState string

const (
	// ShutdownStateRunning indicates the orchestrator is running normally
	ShutdownStateRunning ShutdownState = "running"
	// ShutdownStateInitiated indicates shutdown has been initiated
	ShutdownStateInitiated ShutdownState = "initiated"
	// ShutdownStateStopping indicates subsystems are being stopped
	ShutdownStateStopping ShutdownState = "stopping"
	// ShutdownStateComplete indicates shutdown is complete
	ShutdownStateComplete ShutdownState = "complete"
)

// ShutdownHook is a function that can be called during shutdown
type ShutdownHook func(ctx context.Context) error

// ShutdownManager manages the graceful shutdown process
type ShutdownManager struct {
	mu              sync.RWMutex
	orch            *Orchestrator
	state           ShutdownState
	shutdownTimeout time.Duration
	hooks           []ShutdownHook
	logger          *logger.Logger
	signalChan      chan os.Signal
	shutdownCtx     context.Context
	shutdownCancel  context.CancelFunc
	started         bool
	completionChan  chan struct{}
	shutdownReason  string
}

// NewShutdownManager creates a new shutdown manager
func NewShutdownManager(orch *Orchestrator, timeout time.Duration, log *logger.Logger) *ShutdownManager {
	if log == nil {
		// Create a default logger if none provided - this should rarely happen
		log, _ = logger.NewDefault()
	}

	ctx, cancel := context.WithCancel(context.Background())

	return &ShutdownManager{
		orch:            orch,
		state:           ShutdownStateRunning,
		shutdownTimeout: timeout,
		hooks:           make([]ShutdownHook, 0),
		logger:          log.With("component", "shutdown_manager"),
		signalChan:      make(chan os.Signal, 1),
		shutdownCtx:     ctx,
		shutdownCancel:  cancel,
		started:         false,
		completionChan:  make(chan struct{}),
		shutdownReason:  "",
	}
}

// Start begins listening for shutdown signals
func (sm *ShutdownManager) Start() {
	sm.mu.Lock()
	defer sm.mu.Unlock()

	if sm.started {
		return
	}

	// Register signal handlers for common shutdown signals
	signals := []os.Signal{
		syscall.SIGINT,
		syscall.SIGTERM,
	}
	// SIGQUIT is not available on Windows
	// Add build tag specific signals in platform files if needed
	// syscall.SIGQUIT is available on Unix-like systems

	signal.Notify(sm.signalChan, signals...)

	sm.started = true
	sm.logger.Info("Shutdown manager started",
		"timeout", sm.shutdownTimeout,
		"signals", len(signals))

	// Start signal handler goroutine
	go sm.handleSignals()
}

// Stop stops the shutdown manager (cancels signal handling)
func (sm *ShutdownManager) Stop() {
	sm.mu.Lock()
	defer sm.mu.Unlock()

	if !sm.started {
		return
	}

	signal.Stop(sm.signalChan)
	sm.shutdownCancel()
	sm.started = false

	sm.logger.Debug("Shutdown manager stopped")
}

// Shutdown initiates a graceful shutdown with the specified reason
func (sm *ShutdownManager) Shutdown(ctx context.Context, reason string) error {
	sm.mu.Lock()

	if sm.state != ShutdownStateRunning {
		sm.mu.Unlock()
		return types.NewError(types.ErrCodeFailedPrecondition, "shutdown already initiated")
	}

	sm.state = ShutdownStateInitiated
	sm.shutdownReason = reason
	sm.logger.Info("Shutdown initiated", "reason", reason)
	sm.mu.Unlock()

	// Create shutdown context with timeout
	shutdownCtx, cancel := context.WithTimeout(ctx, sm.shutdownTimeout)
	defer cancel()

	// Execute pre-shutdown hooks
	if err := sm.executeHooks(shutdownCtx, "pre-shutdown"); err != nil {
		sm.logger.Error("Pre-shutdown hooks failed", "error", err)
	}

	// Stop accepting new work
	sm.setState(ShutdownStateStopping)

	// Close the orchestrator
	if err := sm.orch.Close(); err != nil {
		sm.logger.Error("Orchestrator close failed", "error", err)
		// Continue with shutdown anyway
	}

	// Execute post-shutdown hooks
	if err := sm.executeHooks(shutdownCtx, "post-shutdown"); err != nil {
		sm.logger.Error("Post-shutdown hooks failed", "error", err)
	}

	// Mark shutdown as complete
	sm.setState(ShutdownStateComplete)
	close(sm.completionChan)

	sm.logger.Info("Shutdown complete", "reason", reason, "duration", time.Since(sm.getStartTime()))

	return nil
}

// ShutdownAndWait initiates shutdown and waits for completion
func (sm *ShutdownManager) ShutdownAndWait(ctx context.Context, reason string) error {
	// Launch shutdown in goroutine
	errChan := make(chan error, 1)
	go func() {
		errChan <- sm.Shutdown(ctx, reason)
	}()

	// Wait for either completion or context cancellation
	select {
	case err := <-errChan:
		return err
	case <-ctx.Done():
		return types.WrapError(types.ErrCodeCanceled, "shutdown wait canceled", ctx.Err())
	case <-sm.completionChan:
		return nil
	}
}

// AddHook adds a shutdown hook that will be called during shutdown
func (sm *ShutdownManager) AddHook(hook ShutdownHook) {
	sm.mu.Lock()
	defer sm.mu.Unlock()

	sm.hooks = append(sm.hooks, hook)
	sm.logger.Debug("Shutdown hook registered", "total_hooks", len(sm.hooks))
}

// State returns the current shutdown state
func (sm *ShutdownManager) State() ShutdownState {
	sm.mu.RLock()
	defer sm.mu.RUnlock()
	return sm.state
}

// IsShuttingDown returns true if shutdown has been initiated
func (sm *ShutdownManager) IsShuttingDown() bool {
	state := sm.State()
	return state == ShutdownStateInitiated || state == ShutdownStateStopping || state == ShutdownStateComplete
}

// IsComplete returns true if shutdown is complete
func (sm *ShutdownManager) IsComplete() bool {
	return sm.State() == ShutdownStateComplete
}

// ShutdownReason returns the reason for shutdown
func (sm *ShutdownManager) ShutdownReason() string {
	sm.mu.RLock()
	defer sm.mu.RUnlock()
	return sm.shutdownReason
}

// WaitCompletion waits for shutdown to complete
func (sm *ShutdownManager) WaitCompletion(ctx context.Context) error {
	select {
	case <-sm.completionChan:
		return nil
	case <-ctx.Done():
		return types.WrapError(types.ErrCodeCanceled, "wait for completion canceled", ctx.Err())
	}
}

// Context returns the shutdown context
func (sm *ShutdownManager) Context() context.Context {
	return sm.shutdownCtx
}

// handleSignals handles incoming shutdown signals
func (sm *ShutdownManager) handleSignals() {
	for {
		select {
		case sig := <-sm.signalChan:
			reason := fmt.Sprintf("signal received: %s", sig)
			sm.logger.Info("Shutdown signal received", "signal", sig)

			// Initiate shutdown in goroutine to avoid blocking signal handling
			go func() {
				ctx, cancel := context.WithTimeout(context.Background(), sm.shutdownTimeout)
				defer cancel()
				if err := sm.ShutdownAndWait(ctx, reason); err != nil {
					sm.logger.Error("Shutdown failed", "error", err)
				}
			}()

		case <-sm.shutdownCtx.Done():
			sm.logger.Debug("Signal handler stopping")
			return
		}
	}
}

// executeHooks executes all registered shutdown hooks
func (sm *ShutdownManager) executeHooks(ctx context.Context, phase string) error {
	sm.mu.RLock()
	hooks := make([]ShutdownHook, len(sm.hooks))
	copy(hooks, sm.hooks)
	sm.mu.RUnlock()

	sm.logger.Debug("Executing shutdown hooks", "phase", phase, "count", len(hooks))

	var errors []error
	for i, hook := range hooks {
		hookCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
		hookName := fmt.Sprintf("hook-%d", i)

		sm.logger.Debug("Executing shutdown hook", "phase", phase, "hook", hookName)

		func() {
			defer cancel()
			if err := hook(hookCtx); err != nil {
				sm.logger.Error("Shutdown hook failed",
					"phase", phase,
					"hook", hookName,
					"error", err)
				errors = append(errors, err)
			}
		}()

		// Check if context was canceled
		select {
		case <-ctx.Done():
			sm.logger.Warn("Shutdown hook execution canceled", "phase", phase)
			return types.WrapError(types.ErrCodeCanceled, "hook execution canceled", ctx.Err())
		default:
		}
	}

	if len(errors) > 0 {
		return types.WrapError(types.ErrCodePartialFailure, fmt.Sprintf("%s hooks failed", phase), errors[0])
	}

	return nil
}

// setState sets the shutdown state
func (sm *ShutdownManager) setState(state ShutdownState) {
	sm.mu.Lock()
	defer sm.mu.Unlock()
	sm.state = state
	sm.logger.Debug("Shutdown state changed", "state", state)
}

// getStartTime returns the time when shutdown was initiated
func (sm *ShutdownManager) getStartTime() time.Time {
	// This is a simplified version - in production you'd store the actual start time
	return time.Now()
}

// ShutdownGracefully shuts down the orchestrator gracefully with a timeout
func ShutdownGracefully(orch *Orchestrator, timeout time.Duration) error {
	if orch == nil {
		return types.NewError(types.ErrCodeInvalidArgument, "orchestrator is nil")
	}

	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	defer cancel()

	return orch.Close()
}

// ShutdownWithSignalHandling creates a shutdown manager and starts signal handling
func ShutdownWithSignalHandling(orch *Orchestrator, timeout time.Duration, log *logger.Logger) (*ShutdownManager, error) {
	if orch == nil {
		return nil, types.NewError(types.ErrCodeInvalidArgument, "orchestrator is nil")
	}

	sm := NewShutdownManager(orch, timeout, log)
	sm.Start()

	return sm, nil
}

// ShutdownOnContextCancel initiates shutdown when the context is canceled
func ShutdownOnContextCancel(orch *Orchestrator, ctx context.Context, log *logger.Logger) (*ShutdownManager, error) {
	if orch == nil {
		return nil, types.NewError(types.ErrCodeInvalidArgument, "orchestrator is nil")
	}

	sm := NewShutdownManager(orch, 30*time.Second, log)

	go func() {
		<-ctx.Done()
		reason := fmt.Sprintf("context canceled: %v", ctx.Err())
		shutdownCtx, cancel := context.WithTimeout(context.Background(), sm.shutdownTimeout)
		defer cancel()
		if err := sm.ShutdownAndWait(shutdownCtx, reason); err != nil {
			log.Error("Context-based shutdown failed", "error", err)
		}
	}()

	return sm, nil
}

// String returns a string representation of the shutdown state
func (s ShutdownState) String() string {
	return string(s)
}

// String returns a string representation of the shutdown manager
func (sm *ShutdownManager) String() string {
	sm.mu.RLock()
	defer sm.mu.RUnlock()

	return fmt.Sprintf("ShutdownManager{state: %s, timeout: %v, hooks: %d, started: %t}",
		sm.state, sm.shutdownTimeout, len(sm.hooks), sm.started)
}

File: agents/assistant/src/shutdown.ts
// shutdown.ts - Graceful shutdown handler for the Assistant agent
//
// This file handles graceful shutdown of the Assistant agent, including
// signal handling, shutdown hooks, and timeout management.
//
// Copyright 2026 baaaht project

// =============================================================================
// Shutdown State Enumeration
// =============================================================================

/**
 * ShutdownState represents the current state of the shutdown process
 */
export enum ShutdownState {
  /**
   * The agent is running normally
   */
  RUNNING = 'running',

  /**
   * Shutdown has been initiated
   */
  INITIATED = 'initiated',

  /**
   * Subsystems are being stopped
   */
  STOPPING = 'stopping',

  /**
   * Shutdown is complete
   */
  COMPLETE = 'complete',
}

// =============================================================================
// Shutdown Hook Type
// =============================================================================

/**
 * ShutdownHook is a function that can be called during shutdown
 *
 * @param signal - AbortSignal for cancellation
 * @returns Promise that resolves when the hook is complete
 */
export type ShutdownHook = (signal: AbortSignal) => Promise<void>;

// =============================================================================
// Logger Interface
// =============================================================================

/**
 * Logger interface for shutdown logging
 */
export interface ShutdownLogger {
  info(message: string, meta?: Record<string, unknown>): void;
  warn(message: string, meta?: Record<string, unknown>): void;
  error(message: string, meta?: Record<string, unknown>): void;
  debug(message: string, meta?: Record<string, unknown>): void;
}

/**
 * Default console-based logger implementation
 */
class DefaultLogger implements ShutdownLogger {
  private readonly prefix: string;

  constructor(prefix: string = 'ShutdownManager') {
    this.prefix = prefix;
  }

  info(message: string, meta?: Record<string, unknown>): void {
    console.log(`[${this.prefix}] [INFO] ${message}`, meta ? JSON.stringify(meta) : '');
  }

  warn(message: string, meta?: Record<string, unknown>): void {
    console.warn(`[${this.prefix}] [WARN] ${message}`, meta ? JSON.stringify(meta) : '');
  }

  error(message: string, meta?: Record<string, unknown>): void {
    console.error(`[${this.prefix}] [ERROR] ${message}`, meta ? JSON.stringify(meta) : '');
  }

  debug(message: string, meta?: Record<string, unknown>): void {
    if (process.env.DEBUG) {
      console.debug(`[${this.prefix}] [DEBUG] ${message}`, meta ? JSON.stringify(meta) : '');
    }
  }
}

// =============================================================================
// Shutdown Configuration
// =============================================================================

/**
 * ShutdownConfig contains configuration for the shutdown manager
 */
export interface ShutdownConfig {
  /**
   * Shutdown timeout in milliseconds
   * Defaults to 30000 (30 seconds)
   */
  shutdownTimeout?: number;

  /**
   * Hook timeout in milliseconds
   * Defaults to 5000 (5 seconds)
   */
  hookTimeout?: number;

  /**
   * Logger instance
   */
  logger?: ShutdownLogger;
}

// =============================================================================
// Default Configuration
// =============================================================================

const DEFAULT_SHUTDOWN_CONFIG: Required<ShutdownConfig> = {
  shutdownTimeout: 30000, // 30 seconds
  hookTimeout: 5000, // 5 seconds
  logger: new DefaultLogger(),
};

// =============================================================================
// Shutdown Manager Class
// =============================================================================

/**
 * ShutdownManager manages the graceful shutdown process
 *
 * This class handles:
 * - Signal handling (SIGINT, SIGTERM)
 * - Shutdown hook execution (pre-shutdown and post-shutdown)
 * - Timeout management
 * - State tracking
 * - Completion notification
 */
export class ShutdownManager {
  private config: Required<ShutdownConfig>;
  private state: ShutdownState;
  private hooks: ShutdownHook[];
  private logger: ShutdownLogger;

  private started: boolean;
  private completionPromise: Promise<void> | null;
  private completionResolve: (() => void) | null;
  private shutdownReason: string;
  private shutdownStartedAt: Date | null;

  private abortController: AbortController;
  private signalHandlers: {
    sigint: (() => void) | null;
    sigterm: (() => void) | null;
  };

  // Mutex-like lock for preventing concurrent shutdown
  private shutdownLock: boolean;

  /**
   * Creates a new ShutdownManager
   *
   * @param config - Shutdown configuration
   */
  constructor(config: ShutdownConfig = {}) {
    this.config = {
      ...DEFAULT_SHUTDOWN_CONFIG,
      ...config,
    };

    // Use provided logger or default
    this.logger = this.config.logger ?? new DefaultLogger();

    this.state = ShutdownState.RUNNING;
    this.hooks = [];
    this.started = false;
    this.completionPromise = null;
    this.completionResolve = null;
    this.shutdownReason = '';
    this.shutdownStartedAt = null;

    this.abortController = new AbortController();
    this.signalHandlers = {
      sigint: null,
      sigterm: null,
    };

    this.shutdownLock = false;

    this.logger.debug('ShutdownManager created', {
      timeout: this.config.shutdownTimeout,
      hookTimeout: this.config.hookTimeout,
    });
  }

  // ==========================================================================
  // Public Methods - Lifecycle
  // ==========================================================================

  /**
   * Start begins listening for shutdown signals
   */
  start(): void {
    if (this.started) {
      this.logger.debug('Shutdown manager already started');
      return;
    }

    // Register signal handlers for common shutdown signals
    this.signalHandlers.sigint = () => this.handleSignal('SIGINT');
    this.signalHandlers.sigterm = () => this.handleSignal('SIGTERM');

    process.on('SIGINT', this.signalHandlers.sigint);
    process.on('SIGTERM', this.signalHandlers.sigterm);

    this.started = true;

    this.logger.info('Shutdown manager started', {
      timeout: this.config.shutdownTimeout,
      signals: ['SIGINT', 'SIGTERM'],
    });
  }

  /**
   * Stop stops the shutdown manager (cancels signal handling)
   */
  stop(): void {
    if (!this.started) {
      return;
    }

    if (this.signalHandlers.sigint) {
      process.off('SIGINT', this.signalHandlers.sigint);
    }
    if (this.signalHandlers.sigterm) {
      process.off('SIGTERM', this.signalHandlers.sigterm);
    }

    this.abortController.abort();
    this.started = false;

    this.logger.debug('Shutdown manager stopped');
  }

  // ==========================================================================
  // Public Methods - Shutdown
  // ==========================================================================

  /**
   * Shutdown initiates a graceful shutdown with the specified reason
   *
   * @param reason - The reason for shutdown
   * @returns Promise that resolves when shutdown is complete
   */
  async shutdown(reason: string): Promise<void> {
    // Acquire shutdown lock to prevent concurrent shutdowns
    if (this.shutdownLock) {
      // If shutdown is already in progress, wait for it
      if (this.completionPromise) {
        return this.completionPromise;
      }
      return;
    }

    this.shutdownLock = true;

    try {
      // Check if already shutting down
      if (this.state !== ShutdownState.RUNNING) {
        throw new Error(`Shutdown already initiated (state: ${this.state})`);
      }

      this.state = ShutdownState.INITIATED;
      this.shutdownReason = reason;
      this.shutdownStartedAt = new Date();

      this.logger.info('Shutdown initiated', { reason });

      // Create abort signal with timeout
      const abortController = new AbortController();
      const timeoutId = setTimeout(() => {
        abortController.abort();
        this.logger.warn('Shutdown timeout reached', {
          timeout: this.config.shutdownTimeout,
        });
      }, this.config.shutdownTimeout);

      const signal = abortController.signal;

      try {
        // Execute pre-shutdown hooks
        await this.executeHooks(signal, 'pre-shutdown');

        // Transition to stopping state
        this.setState(ShutdownState.STOPPING);

        // Execute post-shutdown hooks
        await this.executeHooks(signal, 'post-shutdown');

        // Mark shutdown as complete
        this.setState(ShutdownState.COMPLETE);

        // Resolve completion promise
        if (this.completionResolve) {
          this.completionResolve();
          this.completionResolve = null;
        }
        this.completionPromise = null;

        const duration = this.shutdownStartedAt
          ? Date.now() - this.shutdownStartedAt.getTime()
          : 0;

        this.logger.info('Shutdown complete', {
          reason,
          durationMs: duration,
        });
      } finally {
        clearTimeout(timeoutId);
      }
    } catch (err) {
      const error = err instanceof Error ? err : new Error(String(err));
      this.logger.error('Shutdown failed', { error: error.message });
      throw error;
    } finally {
      // Release lock
      this.shutdownLock = false;
    }
  }

  /**
   * ShutdownAndWait initiates shutdown and waits for completion
   *
   * @param reason - The reason for shutdown
   * @param timeout - Optional timeout in milliseconds
   * @returns Promise that resolves when shutdown is complete or rejects on timeout
   */
  async shutdownAndWait(reason: string, timeout?: number): Promise<void> {
    // Launch shutdown in background
    const shutdownPromise = this.shutdown(reason);

    // Wait for either completion or timeout
    if (timeout) {
      await Promise.race([
        shutdownPromise,
        new Promise<void>((_, reject) =>
          setTimeout(() => reject(new Error(`Shutdown wait timeout after ${timeout}ms`)), timeout)
        ),
      ]);
    } else {
      await shutdownPromise;
    }
  }

  // ==========================================================================
  // Public Methods - Hooks
  // ==========================================================================

  /**
   * AddHook adds a shutdown hook that will be called during shutdown
   *
   * @param hook - The shutdown hook to add
   */
  addHook(hook: ShutdownHook): void {
    this.hooks.push(hook);
    this.logger.debug('Shutdown hook registered', {
      totalHooks: this.hooks.length,
    });
  }

  /**
   * RemoveHook removes a shutdown hook
   *
   * @param hook - The shutdown hook to remove
   * @returns True if the hook was removed, false otherwise
   */
  removeHook(hook: ShutdownHook): boolean {
    const index = this.hooks.indexOf(hook);
    if (index !== -1) {
      this.hooks.splice(index, 1);
      this.logger.debug('Shutdown hook removed', {
        totalHooks: this.hooks.length,
      });
      return true;
    }
    return false;
  }

  // ==========================================================================
  // Public Methods - State Query
  // ==========================================================================

  /**
   * State returns the current shutdown state
   */
  getState(): ShutdownState {
    return this.state;
  }

  /**
   * IsShuttingDown returns true if shutdown has been initiated
   */
  isShuttingDown(): boolean {
    return (
      this.state === ShutdownState.INITIATED ||
      this.state === ShutdownState.STOPPING ||
      this.state === ShutdownState.COMPLETE
    );
  }

  /**
   * IsComplete returns true if shutdown is complete
   */
  isComplete(): boolean {
    return this.state === ShutdownState.COMPLETE;
  }

  /**
   * ShutdownReason returns the reason for shutdown
   */
  getShutdownReason(): string {
    return this.shutdownReason;
  }

  /**
   * GetSignal returns the abort signal for cancellation
   */
  getSignal(): AbortSignal {
    return this.abortController.signal;
  }

  /**
   * WaitCompletion waits for shutdown to complete
   *
   * @param timeout - Optional timeout in milliseconds
   * @returns Promise that resolves when shutdown is complete
   */
  async waitCompletion(timeout?: number): Promise<void> {
    if (this.isComplete()) {
      return;
    }

    // Create completion promise if not exists
    if (!this.completionPromise) {
      this.completionPromise = new Promise<void>((resolve) => {
        this.completionResolve = resolve;
      });
    }

    if (timeout) {
      await Promise.race([
        this.completionPromise,
        new Promise<void>((_, reject) =>
          setTimeout(() => reject(new Error(`Wait completion timeout after ${timeout}ms`)), timeout)
        ),
      ]);
    } else {
      await this.completionPromise;
    }
  }

  // ==========================================================================
  // Private Methods - Signal Handling
  // ==========================================================================

  /**
   * Handle an incoming shutdown signal
   *
   * @private
   */
  private handleSignal(signalName: string): void {
    const reason = `signal received: ${signalName}`;
    this.logger.info('Shutdown signal received', { signal: signalName });

    // Initiate shutdown in background to avoid blocking signal handling
    this.shutdownAndWait(reason, this.config.shutdownTimeout).catch((err) => {
      this.logger.error('Shutdown from signal failed', {
        error: err instanceof Error ? err.message : String(err),
      });
    });
  }

  // ==========================================================================
  // Private Methods - Hook Execution
  // ==========================================================================

  /**
   * Execute all registered shutdown hooks
   *
   * @private
   * @param signal - AbortSignal for cancellation
   * @param phase - The shutdown phase (pre-shutdown or post-shutdown)
   */
  private async executeHooks(signal: AbortSignal, phase: string): Promise<void> {
    // Copy hooks to avoid modification during iteration
    const hooks = [...this.hooks];

    this.logger.debug('Executing shutdown hooks', {
      phase,
      count: hooks.length,
    });

    const errors: unknown[] = [];

    for (let i = 0; i < hooks.length; i++) {
      const hook = hooks[i];
      const hookName = `${phase}-hook-${i}`;

      // Check if already aborted
      if (signal.aborted) {
        this.logger.warn('Shutdown hook execution canceled', { phase });
        throw new Error('Shutdown aborted during hook execution');
      }

      this.logger.debug('Executing shutdown hook', { phase, hook: hookName });

      try {
        // Execute hook with timeout
        await Promise.race([
          hook(signal),
          new Promise<void>((_, reject) =>
            setTimeout(() => reject(new Error(`Hook timeout after ${this.config.hookTimeout}ms`)), this.config.hookTimeout)
          ),
        ]);
      } catch (err) {
        const error = err instanceof Error ? err : new Error(String(err));
        this.logger.error('Shutdown hook failed', {
          phase,
          hook: hookName,
          error: error.message,
        });
        errors.push(error);
      }

      // Check if aborted during hook execution
      if (signal.aborted) {
        this.logger.warn('Shutdown hook execution canceled', { phase });
        throw new Error('Shutdown aborted during hook execution');
      }
    }

    if (errors.length > 0) {
      throw new Error(
        `${phase} hooks failed: ${errors.map((e) => (e instanceof Error ? e.message : String(e))).join(', ')}`
      );
    }
  }

  // ==========================================================================
  // Private Methods - State Management
  // ==========================================================================

  /**
   * Set the shutdown state
   *
   * @private
   */
  private setState(state: ShutdownState): void {
    const previousState = this.state;
    this.state = state;
    this.logger.debug('Shutdown state changed', {
      from: previousState,
      to: state,
    });
  }

  // ==========================================================================
  // Public Methods - String Representation
  // ==========================================================================

  /**
   * String returns a string representation of the shutdown manager
   */
  toString(): string {
    return `ShutdownManager{state: ${this.state}, timeout: ${this.config.shutdownTimeout}ms, hooks: ${this.hooks.length}, started: ${this.started}}`;
  }
}

// =============================================================================
// Factory Functions
// =============================================================================

/**
 * Create a new ShutdownManager instance
 *
 * @param config - Shutdown configuration
 * @returns A new ShutdownManager instance
 */
export function createShutdownManager(config?: ShutdownConfig): ShutdownManager {
  return new ShutdownManager(config);
}

// =============================================================================
// Re-exports
// =============================================================================
